File: Python/Problems/Problems/problem628.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem


class Problem628(Problem):
    def test(self) -> bool:
        nums = [1, 2, 3]

        expected = 6

        result = self.maximumProduct(nums)

        print(result)

        return result == expected

    def maximumProduct(self, nums: List[int]) -> int:
        max1 = -sys.maxsize
        max2 = -sys.maxsize
        max3 = -sys.maxsize
        min1 = sys.maxsize
        min2 = sys.maxsize
        for n in nums:
            if n >= max1:
                max3 = max2
                max2 = max1
                max1 = n
            elif n >= max2:
                max3 = max2
                max2 = n
            elif n > max3:
                max3 = n

            if n <= min1:
                min2 = min1
                min1 = n
            elif n < min2:
                min2 = n
        return max(min1 * min2 * max1, max1 * max2 * max3)

File: C++/Problems/Problems/problem_628.cpp
#include "problem_628.h"

bool problem_628::test() {
    std::vector nums { 1, 2, 3 };

    int expected = 6;

    auto result = maximumProduct(nums);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_628::maximumProduct(std::vector<int> &nums) {
    int max1 = INT_MIN;
    int max2 = INT_MIN;
    int max3 = INT_MIN;
    int min1 = INT_MAX;
    int min2 = INT_MAX;
    for (auto n : nums)
    {
        if (n >= max1)
        {
            max3 = max2;
            max2 = max1;
            max1 = n;
        }
        else if (n >= max2)
        {
            max3 = max2;
            max2 = n;
        }
        else if (n > max3)
            max3 = n;

        if (n <= min1)
        {
            min2 = min1;
            min1 = n;
        }
        else if (n < min2)
            min2 = n;
    }
    return std::max(min1 * min2 * max1, max1 * max2 * max3);
}
